File: 0101-symmetric-tree/0101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        return self.isMirror(root.left, root.right)
        
    def isMirror(self, left, right):
        # 둘다 null
        if not left and not right:
            return True
        # 한쪽만 null
        elif not left or not right:
            return False
        
        return left.val == right.val and self.isMirror(left.left, right.right) and self.isMirror(left.right, right.left)

File: 0101-symmetric-tree/0101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        // 최상단이 null 이면 true
        if (root == null) {
            return true;
        }
        // 왼쪽 노드와 오른쪽 노드 비교하기
        return isMirror(root.left, root.right);
    }
    
    private boolean isMirror(TreeNode node1, TreeNode node2) {
        // 둘다 null 이면 True
        if (node1 == null && node2 == null) {
            return true;
        }
        // 한쪽만 null 이면 False
        else if (node1 == null || node2 == null) {
            return false;
        }
        // 왼쪽  오른쪽 값 비교 그리고 다음 노드들의 대칭 비교 결과
        return node1.val == node2.val && isMirror(node1.left, node2.right) && isMirror(node1.right, node2.left);
    }
}